### DISCUSSION TITLE
Where is the protocol at and where do we need it to be?

### DISCUSSION THREAD

**Discussion by mcsadmin** [CredComSoc/ProtocolWorkingGroup#3] [General]
This is (will!) probably a great place to start if you are new to this Group.

**Comment by JohnG-cc** [comment]
I thought it might be a good idea to take advantage of [Cunningham's Law](https://meta.wikimedia.org/wiki/Cunningham%27s_Law) to kick of discussion by posting my current understanding of the state of the [Credit Commons Protocol](https://gitlab.com/credit-commons) and what still needs doing.

### My understanding of the current state of the protocol is that:

A. There is a complete definition of the [Accounting Protocol](https://creditcommons.net/assets/credit-commons.pdf) that was developed by @matslats and Tim Jenkins.

B. @matslats has created a reference implementation / proof of concept [on GitLab](https://gitlab.com/credit-commons). This implementation is complete and functional, [as can be seen in a demo walkthrough](https://www.youtube.com/watch?v=uQ1EYC8iWMw).

C. There is [a Network Protocol that is implemented as a RESTful Web Service API](https://gitlab.com/credit-commons/cc-php-lib/-/blob/master/docs/credit-commons-v0.2.openapi3.yml), i.e. the protocol definition consists of JSON messages and particular HTTP endpoints that a conformant server should make available.

D. There are two GitLab projects that provide language-specific (PHP) libraries that software developers can include in their PHP projects to make use of the Credit Commons Protocol and to send and receive credit commons transactions across the tree. These are [The Credit Commons PHP Library](https://gitlab.com/credit-commons/cc-php-lib) and the [Credit Commons Node](https://gitlab.com/credit-commons/cc-node). 

E. There is a GitLab project for a [PHP API server application](https://gitlab.com/credit-commons/cc-server) that uses the aforementioned PHP library and node to implement a stand-alone server, that client applications can call using a HTTP REST interface instead of having to embed the node and PHP library. This would be useful for example to a LETS application writen in Javascript, that could call the server as a client to initiate transitive credit commons transactions across the tree.

F. The reference implementation uses a microservice architecture to indirect access to Account Stores and a Business Rules Engine via JSON web service. This would be useful for systems that are written in other languages, such as our hypothetical Javascript LETS application above, to be able to implement it's own Account Store and and Business Rules Engine in Javascript, and have the stand-alone PHP REST server call these - i.e. they get their own authorisation and business rules without having to write PHP. The PHP library includes some dummy implementations of both Account Service and Business Logic Service, however these are not suitable for production use.

G. The documentation for the protocol and API implementation is distributed across the white paper, the various GitLab project READMEs, and files within the project tree (most notably the PHP library). There is also [a separate GitLab project for documentation](), which notes in part that "The Protocol is described here in functional terms in plain english, here as a microservice architecture and implemented here as an API. Note that these may not agree in detail at all points." - i.e. there may be work to do to synchronise the code and documentation from the varous projects. 

H. There is [a demo client](https://gitlab.com/credit-commons/cc-dev-client) that can be used for demonstration and testing purposes, that also acts as a reference example of how to call and work with the HTTP server.

I. There are currently no production implementations of the protocol.

### My understanding of what we need the protocol to be and what needs to be done to achieve that:

In terms of Direction (#2), we want the Credit Commons Protocol to be widely implemented in many Mutual Credit systems worldwide, with many millions of users.

In terms of Vision, this means the protocol will have to be well proven in production use, and clearly documented for both users and software developers. There should also be standard libraries available in the most popular web and app  development languages. We will need to address scalability and availability concerns, either by ensuring we either a) architect a single worldwide Trunk that can handle every transaction at once, will never become unavailable, and cannot be DDOSed, hacked, blocked, or confiscated, or b) provide an additional layer on top of the protocol that allows for a "Forest" of interlocking Credit Commons trees that provide redundant network paths, without devolving into a mesh. We should also ensure that the protocol will be able to work with current and future non-web transport mechanisms, such as peer-to-peer bluetooth networking, email, sms, etc. 

In terms of Next Steps, I understand that:

1. @matslats is currently working to integrate the current protocol and implementation into a Drupal rewrite of his CIC LETS software, the current version of which has around 20K users.

2. @dilgreen is in discussion with someone from the conference who has LETS system with a large number of groups, who is interested in using a to-be-developed Mutual Credit software system and would like that system to integrate with Credit Commons.

3. The GitLab documentation project states that "A Proof-of-concept Protocol and reference server implementation were created in 2020 (now deprecated)" and that a beta release (which links to the repositories discussed above) is in active development. I understand from @matslats however that this effort is now complete.

4. The documentation project also states that the project's needs are a "white hat" audit for robustness and presumably security, coding help with the PHP projects (presumably complete now?), coding help with a reference client (presumably complete now?), improvement of the repo structure, and better diagrams. It also notes that the documentation is itself under development, and I presume this is still the case?

### Ideas for what to do next

So based on all of that, I think I understand the current task of this working group to be to achieve the "productionisation" of the current reference implementation, including reorganising the repositories if necessary, and collating all of the disparate sources of documentation into a singular "Users and Developers Guide" that can be a single up-to-date source of truth for the protocol and project. We should then guide and assist with the implementation of the protocol into it's first production implementations, being @matslats CIC rewrite - and I guess this is mostly about just coordinating any changes to protocols and libraries, and making sure we don't step on each other's toes with breaking changes, and also the potential upcoming development of a Mutual Credit system by/for/with @dilgreen's contact.

. . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 
 
I am sure there are plenty of misconceptions in all of that, so your corrections would be greatly appreciated!

**Comment by matslats** [comment]
I feel heard!
Very few minor comments, not changes, even.
Documentation is always in progress. I believe what's there reflects pretty closely what is, although I would drop the microservices language.
I imagine very rare cases when an AccountStore would be a standalone app. It is intended as the simplest possible extension to the main app to expose the accounts and balance limits needed by the ledger.
We're about to upgrade clearingcentral.net to Credit Commons. This will be a production use-case, but most access will be done through a wrapper which implements the old API, so that the ces servers can upgrade in their own time.

**Comment by JohnG-cc** [comment]
@matslats I'm glad to hear I'm on the right track with it all!

Good to know that the documentation is current. Do you think it would be a good idea if it the documentation project was expanded to include all the documentation from the other repos too, for example the contents of the READMEs and the files in cc-php-lib/documentation too?

Should "microservices" references be dropped because it's really just the business engine and AccountStore that are services, and the AccountStore not usually at that? Or are you rethinking the architecture?

My current understanding is that the distinction between a bit of code being in cc-php-lib or node is that things needed by nodes and clients goes in cc-php-lib, and things only needed by nodes and servers goes in node, and of course the server and client are their own repos... is that right?

I was surprised to learn from @utunga yesterday that SvenskBarter is using an instance of the server as it's back end ledger! @dilgreen probably mentioned this to me already but I wouldn't have understood the significance at the time. I don't think I had fully understood the implications of the node handling balances in the BoT account - this of course means that the node is handling all the *other* accounts in the ledger too, so has to be a fully fledged ledger system if it stands alone. If some group have their own ledger, they would expose that as an AccountStore, and the node/server could access that, yes?

Given that my summary of "where it needs to be" didn't raise any red flags, I'll start some other discussions around particular parts of that, so we can have a focused discussion on each one.

**Reply by matslats** [reply]
A previous opinion was that many different kinds of documentation were required, So what exists now is a partial attempt to start building that out. It feels a bit diverse and messy to me though, and I'd be glad if someone experienced were to take control.
The first draft was consciously attempting to break things down into microservices, but I folded them into cc-node because it was getting clumsy. The remaining services, AccountStore and BLogic are separate because they need to be.
Yes the cc-php-lib attempts to be the code that is needed by leafs AND nodes, which means mostly object interfaces with very basic classes and the NodeRequester.
Not sure what would have surprised you. The cc-node is a full mutual credit ledger, and the protocol allows a node to act as an account on another ledger. The protocol is the same whether you are a mobile phone connected to your local node, or a country node connected to an international node.
The accounts are not part of cc-node but accessed via a microservice, i.e. the main application must implement a very simple RESTAPI and provide balance limits for each account. The only data that cc-node stores is transaction data, and request/response log.

**Reply by JohnG-cc** [reply]
I don't know about experienced, but I'd be happy to work on collating it into a central and easily navigated set of documents (in the existing doc repo probably?) if you like.

I had wondered where all the microservices were, that explains it. I think that was a good call, the system boundaries seem to be in the right place to me. 

I was surprised because I had somehow formed the impression that having embeddable libraries somehow implied that the inter-ledger feature that was the defining capability of the Credit Commons Protocol could be mixed in to an existing LETS system that had it's own ledger - but of course it's obvious on reflection that could  not work. It felt like an easy mistake to make though, perhaps I should add something to the documentation to preemptively clarify that point?

**Comment by matslats** [comment]
Not sure the best way for you to submit documentation. 
Would you create your own fork and then commit a change and do a pull request.
Seems very cumbersome.
For now I'm happy if you just offer a block of test and a location for it, then I can review it as well.

**Comment by JohnG-cc** [comment]
The first thing I planned to do was to literally collate the text you have written into a single repo and to put give the documents a navigation sidebar, so it was easy to see at a glance what all the documentation is and to jump to the section that one needs. 

This would still involve multiple files, and code changes to specific files, so I think doing that in a cloned repository might be a good way - then you could look at it in-place in the clone and see what you think?

**Comment by matslats** [comment]
This approach sounds as good as any, but please confer with @dilgreen regarding the documentation plan we had started to follow before starting fresh.

**Comment by JohnG-cc** [comment]
@dilgreen, what do you think about collating the docs together? Is there a good reason to keep them distributed?

**Reply by dilgreen** [reply]
The previous attempt at integration hardly started before I ran out of time to allocate to it, so I wouldn't worry about it. The intent was to use the gitlab equivalent of github pages.
As I understand it, you suggest a simple approach in the first instance - a single document with a TOC.
This sounds good - relatively speedy and definitely collecting everything on one place.

**Reply by JohnG-cc** [reply]
If no one has any objections then, I might start picking away at that, and solicit comments when there is something to see.
